docs/proposals/20190704-AutomatedUpgrades.md

```diff
@@ -0,0 +1,206 @@
+This provides a clean slate for upgrades.
+
+> TODO: Exactly how `machined` is executed needs to be figured out.
+> We could `switch_root` or we could not.
+
+#### The Scope of `machined`
+
+Once we enter the `rootfs` we have three high level tasks:
+
+- retreive the machineconfig
+- create, format, and mount partitions per the builtin specifications
+- start system and k8s.io services
+
+By having a versioned `machined` that is decoupled from `initramfs` we can encapsulate a version of Talos with one binary that knows how to run Talos.
+It will know how to handle versioned machineconfigs and migrate them to newer formats.
+If we ever decide to change paritition layout, filesystem types, or anything at the blockdevice level, the versioned `machined` will know how to handle that too.
+The same can be said about the services that `machined` is responsible for.
+
+### Running `rootfs` in Memory
+
+To take nodes one step closer to becoming completely ephemeral, we will run the `rootfs` in memory.
+This is part of the solution to remove the need for `ROOT-A` and `ROOT-B` partitions.
+It also allows us to dedicate the owned block device to data that can be reproduced (e.g. containerd data, kubelet data, etcd data, CNI data, etc.).
+
+### Using `squashfs` for the `rootfs`
+
+For performance reasons, and a smaller footprint, we can create our `rootfs` as a `squashfs` image.
+The way in which we can run the `rootfs` is by mounting it using a loopback device.
+Once mounted, we have two options:
+
+- perform a `switch_root` into the mounted `rootfs`
+- `fork` a `machined` process without performing a `switch_root`
+
+Another benefit of `squashfs` is that it allows us to retain our read only `rootfs` without having to mount it as such.
+
+> TODO: More research needs to be done here.
+> The driving force behind this decision is figuring out how we can return control back to `init` in order to reset a node.
+
+### Slimming down `rootfs`
+
+#### Proposal 1
+
+Today our `rootfs` contains large tarballs from Kubernetes.
+In this implementation we will remove these tarballs to a dedicated `cache` tarball.
+This allows us to considerably reduce the size of the `rootfs`.
+
+The downside to this is that the tarballs are no longer on a read only partition and can be overwritten, taking us a step away from our immutability goals.
+
+#### Proposal 2
+
+An alternative approach is to use compression.
+We currently do not compress the container tarballs.
+The idea here is that by compressing the container images, and then creating a `squashfs`, we can reduce the `rootfs` down to an acceptable size.
+This has the advantage that everything is nicely bundled together.
+
+A downside is that it will take the kernel longer to decompress the `initramfs`.
+Another downside is that we would require at least 1GB of RAM to run talos.
+
+#### Proposal 3
+
+Finally, we could also remove the Kubernetes tarballs entirely and let the `kubelet` download them.
```

Comment:
can we download approved tarball from some Talos location? like customer might override the URL, but we verify the checksum?